sdks/python/apache_beam/typehints/typecheck.py

```diff
@@ -265,3 +268,89 @@ def visit_transform(self, applied_transform):
                 transform.get_type_hints(),
                 applied_transform.full_label),
             applied_transform.full_label)
+
+
+class PerformanceTypeCheckVisitor(pipeline.PipelineVisitor):
+
+  _in_combine = False
+  combine_classes = (
+      core.CombineFn,
+      core.CombinePerKey,
+      core.CombineValuesDoFn,
+      core.CombineValues,
+      core.CombineGlobally)
+
+  def enter_composite_transform(self, applied_transform):
+    if isinstance(applied_transform.transform, self.combine_classes):
+      self._in_combine = True
+
+  def leave_composite_transform(self, applied_transform):
+    if isinstance(applied_transform.transform, self.combine_classes):
+      self._in_combine = False
+
+  def visit_transform(self, applied_transform):
+    transform = applied_transform.transform
+    if isinstance(transform, core.ParDo) and not self._in_combine:
+      # Prefix label with 'ParDo' if necessary
+      full_label = applied_transform.full_label
+      if not full_label.startswith('ParDo'):
+        full_label = 'ParDo(%s)' % full_label
+
+      # Store output type hints in current transform
+      transform.fn._runtime_output_constraints = {}
```

Comment:
This will overwrite any typehint constraints that may have been set earlier. Instead, use the same method.